greedy_max/schedule.py

```diff
@@ -1,65 +1,125 @@
-                            self.disperser != 'null' and \
-                            self.instrument != 'GMOS' else 0.0 * u.min
-            # save to total time or add units
-            return int(np.ceil(tcal.to(u.h) / dt.to(u.h))) 
-        return 0
+    def length(self) -> int:
+        """
+        Calculate the length of the unit based on both observation and calibrations times
+        """
+        obs_slots = sum([obs.length for obs in self.observations])
 
+        if self.standard_time > 0: # not a single science observation
+            standards_needed = max(1, int(obs_slots // self.standard_time))
 
-# TODO: Implement scheduling groups
-# class SchedulingGroup:
-#    def __init__():
+            if standards_needed == 1:
+                cal_slots = self.calibrations[0].length #take just one
+            else:
+                cal_slots = sum([cal.length for cal in self.calibrations])
 
+            return obs_slots + cal_slots
+        else:
+            return obs_slots
+
+    def observed(self) -> int:
+        """
+        Calculate the observed time for both observation and calibrations
+        """
+        obs_slots = sum([obs.observed for obs in self.observations])
+        cal_slots = sum([cal.observed for cal in self.calibrations])
+        return obs_slots + cal_slots
+    
+    def acquisition(self) -> None:
+        """
+        Add acquisition overhead to the total length of each observation in the unit
+        """
+        for observation in self.observations:
+            if observation.observed < observation.length: # not complete observation
+                observation.length += observation.acquisition
+
+    def get_observations(self) -> Dict[int,Observation]:
+        total_obs = {}
+        for obs in self.observations:
+            total_obs[obs.idx] = obs
+        for cal in self.calibrations:
+            total_obs[cal.idx] = cal
+        return total_obs
+    
+    def __contains__(self, obs_idx:int) -> bool:
+        
+        if obs_idx in [sci.idx for sci in self.observations]:
+            return True
+        elif obs_idx in [cal.idx for cal in self.calibrations]:
+            return True
+        else:
+            return False
+    
+    def __str__(self) -> str:
+        return f'Unit {self.idx} \n\
+                 -- observations: {[sci.idx for sci in self.observations]} \n\
+                 -- calibrations: {[cal.idx for cal in self.calibrations]}'
+    
+    
 
 class TimeSlots:
     decoder = {'A':'0','B':'1','Q':'0',
                 'C':'1','LP':'2','FT':'3',
                 'SV':'8','DD':'9'}
-    #pattern = '|'.join(map(re.escape, decoder.keys()))
     pattern = '|'.join(decoder.keys())
     def __init__(self, 
-                 time_slot_length: Quantity, 
-                 weights: List[int] , 
+                 time_slot_length: float, 
+                 weights: Dict[Site,List[int]],
+                 airmass:  Dict[Site,List[int]], 
```

Comment:
Is `airmass` really an `int`? (This is a personal question for my own understanding.)

Reply:
Yes, the correct type is  `Dict[Site,Dict[ int, List[float]]]` which is quite long. So I create the type `ValuesByObservation = Dict[int,List[float]]` so both weights and airmass are: 
```
 weights: Dict[Site,ValuesByObservation],
 airmass:  Dict[Site,ValuesByObservation], 
```